### ISSUE TITLE
Crash in sort() with Union element type

### ISSUE THREAD

**Issue by nalimilan** [JuliaLang/julia#23764]
This is fully reproducible on master with CategoricalArrays and Nulls. Also happens with `--check-bounds=yes`. It works fine on 0.6.
```julia
julia> using CategoricalArrays, Nulls

julia> x = CategoricalArray{Union{Null, Int}}([1, 2])
2-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, Int64},1,UInt32}:
 CategoricalArrays.CategoricalValue{Int64,UInt32} 1
 CategoricalArrays.CategoricalValue{Int64,UInt32} 2

julia> sort(x)
Unreachable reached at 0x7fe2f43186a6

signal (4): Illegal instruction
while loading no file, in expression starting on line 0
#sort#8 at ./sort.jl:674
jl_call_fptr_internal at /home/milan/Dev/julia/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1956
jl_apply at /home/milan/Dev/julia/src/julia.h:1449 [inlined]
jl_invoke at /home/milan/Dev/julia/src/gf.c:51
sort at ./sort.jl:674
jl_call_fptr_internal at /home/milan/Dev/julia/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1956
do_call at /home/milan/Dev/julia/src/interpreter.c:70
eval at /home/milan/Dev/julia/src/interpreter.c:262
jl_interpret_toplevel_expr_in at /home/milan/Dev/julia/src/interpreter.c:50
jl_toplevel_eval_flex at /home/milan/Dev/julia/src/toplevel.c:607
jl_toplevel_eval_in at /home/milan/Dev/julia/src/builtins.c:529
eval at ./repl/REPL.jl:3
jl_call_fptr_internal at /home/milan/Dev/julia/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1956
eval_user_input at ./repl/REPL.jl:69
jl_call_fptr_internal at /home/milan/Dev/julia/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1956
macro expansion at ./repl/REPL.jl:100 [inlined]
#1 at ./event.jl:96
jl_call_fptr_internal at /home/milan/Dev/julia/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1956
jl_apply at /home/milan/Dev/julia/src/julia.h:1449 [inlined]
start_task at /home/milan/Dev/julia/src/task.c:268
unknown function (ip: 0xffffffffffffffff)
Allocations: 22252853 (Pool: 22251537; Big: 1316); GC: 43
Illegal instruction (core dumped)
```

**Comment by quinnj** [comment]
I cant' even load CategoricalArrays on master due to https://github.com/JuliaLang/julia/issues/23456

**Comment by quinnj** [comment]
Ok, working around #23456, here's the minimum example that _doesn't_ crash
```julia
ulia> using CategoricalArrays, Nulls

julia> x = CategoricalArray{Union{Null, Int}}([1, 2])
2-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, Int64},1,UInt32,Int64,Nulls.Null}:
 CategoricalArrays.CategoricalValue{Int64,UInt32} 1
 CategoricalArrays.CategoricalValue{Int64,UInt32} 2

julia> sort!(x)
2-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, Int64},1,UInt32,Int64,Nulls.Null}:
 CategoricalArrays.CategoricalValue{Int64,UInt32} 1
 CategoricalArrays.CategoricalValue{Int64,UInt32} 2
```
i.e. we have this definition in sort.jl
```julia
sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...)
```
and calling `sort!` directly results in no crash. I can even do
```julia
julia> sort!(Base.copymutable(x))
2-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, Int64},1,UInt32,Int64,Nulls.Null}:
 CategoricalArrays.CategoricalValue{Int64,UInt32} 1
 CategoricalArrays.CategoricalValue{Int64,UInt32} 2
```
and no crash.

I'm going to dig more into `Base.copymutable`, but could this be some kind of keyword arg issue?

**Comment by quinnj** [comment]
Hmmm, @nalimilan, could this have something to do w/ how you're handling `Union{Null, T}` element types? I ask because I'm seeing
```julia
julia> using CategoricalArrays, Nulls

julia> x = CategoricalArray{String}(["1", "2"])
2-element CategoricalArrays.CategoricalArray{String,1,UInt32,String,Union{}}:
 CategoricalArrays.CategoricalValue{String,UInt32} "1"
 CategoricalArrays.CategoricalValue{String,UInt32} "2"

julia> # sort!(Base.copymutable(x))

       sort(x)
2-element CategoricalArrays.CategoricalArray{String,1,UInt32,String,Union{}}:
 CategoricalArrays.CategoricalValue{String,UInt32} "1"
 CategoricalArrays.CategoricalValue{String,UInt32} "2"

julia> x = CategoricalArray{Union{Void, String}}(["1", "2"])
2-element CategoricalArrays.CategoricalArray{Union{Void, String},1,UInt32,Union{Void, String},Union{}}:
 CategoricalArrays.CategoricalValue{Union{Void, String},UInt32} "1"
 CategoricalArrays.CategoricalValue{Union{Void, String},UInt32} "2"

julia> # sort!(Base.copymutable(x))

       sort(x)
2-element CategoricalArrays.CategoricalArray{Union{Void, String},1,UInt32,Union{Void, String},Union{}}:
 CategoricalArrays.CategoricalValue{Union{Void, String},UInt32} "1"
 CategoricalArrays.CategoricalValue{Union{Void, String},UInt32} "2"

julia> x = CategoricalArray{Union{Null, String}}(["1", "2"])
2-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, String},1,UInt32,String,Nulls.Null}:
 CategoricalArrays.CategoricalValue{String,UInt32} "1"
 CategoricalArrays.CategoricalValue{String,UInt32} "2"

julia> # sort!(Base.copymutable(x))

       sort(x)
Unreachable reached at 0x10f8dd4b6

signal (4): Illegal instruction: 4
while loading no file, in expression starting on line 0
#sort#8 at ./sort.jl:674
```
i.e. it only crashes when `Null` is used _specifically_, and not, for example, if it's `Union{Void, T}`.

**Comment by nalimilan** [comment]
Yes, there's some special handling of `Null` [in the inner constructor](https://github.com/JuliaData/CategoricalArrays.jl/blob/cff2b8b9bc83c3832c1afbaf42c333d59e5bf3f7/src/typedefs.jl#L52). As you can see, you get `CategoricalValue{Union{String, Void}}` if you pass `Union{String, Void}`, but `Union{CategoricalValue{String}, Null}}` if you pass `Union{String, Null}`. So the code paths are quite different. Then you also have specific checks in many methods.

**Comment by vtjnash** [comment]
MRE:
```julia
julia> struct AArray{T, N, S} <: AbstractArray{Union{T, S}, N}; end

julia> typeintersect(Tuple{AArray{Int, 1, T} where T}, Tuple{AbstractArray{T,N}} where {T,N})
Union{}
```

**Comment by vtjnash** [comment]
Seems like this might be related to (but not fixed by) #23700.

**Comment by JeffBezanson** [comment]
I managed to fix that example, but then there is this harder one:

```
julia> struct A{T, N, S} <: AbstractArray{Union{Ref{T}, S}, N}; end

julia> Tuple{A{T, 1, Void} where T} <: Tuple{AbstractArray{T,N} where {T,N}}
true

julia> Tuple{A{T, 1, Void} where T} <: Tuple{AbstractArray{T,N}} where {T,N}
false
```

This seems to be a case where `Ref{T}` (with `T` free) appears in a derived var bound, yet the answer is supposed to be `true`.
